GraphCacheConfig not compatible with custom Scalars

Using a custom Date scalar seems to be incompatible with the current type definitions, since it's neither a Primitive nor a ScalarObject.

```ts
export declare type Primitive = null | number | boolean | string;
export interface ScalarObject {
    [key: string]: unknown;
}
export declare type Scalar = Primitive | ScalarObject;
```


urql version & exchanges:
```json
    "@graphql-codegen/typescript-urql-graphcache": "^1.0.1",
    "@urql/core": "^2.1.1",
    "@urql/exchange-graphcache": "^4.1.0"
```
Also using a custom scalar exchange, similar to https://github.com/clentfort/urql-custom-scalars-exchange to actually convert the scalar values to the correct types.

**Steps to reproduce**

```yml
config:
  scalars:
    DateTime: Date
```

**Expected behavior**

Should accept custom scalars

**Actual behavior**

```sh
error TS2344: Type 'GraphCacheConfig' does not satisfy the constraint 'Partial<CacheExchangeOpts>'.
  Types of property 'resolvers' are incompatible.
    Type 'GraphCacheResolvers' is not assignable to type 'ResolverConfig'.
      Property 'Exam' is incompatible with index signature.
        Type '{ id?: Resolver<WithTypename<Foo>, null, string>; name?: Resolver<WithTypename<Foo>, null, string>; description?: Resolver<...>; ... 25 more ...; canUpdate?: Resolver<...>; }' is not assignable to type '{ [fieldName: string]: Resolver<DataFields, Variables, ResolverResult>; }'.
          Property 'id' is incompatible with index signature.
            Type 'Resolver<WithTypename<Foo>, null, string>' is not assignable to type 'Resolver<DataFields, Variables, ResolverResult>'.
              Type 'WithTypename<Foo>' is not assignable to type 'DataFields'.
                Property 'createdAt' is incompatible with index signature.
                  Type 'Date' is not assignable to type 'DataField'. <<<<<<
                    Type 'Date' is not assignable to type 'ScalarObject'.
                      Index signature is missing in type 'Date'.
```

<!--
  Your issue will be addressed much faster if you include a
  reproduction. Here are some templates to help you create one:
  - client: https://codesandbox.io/s/urql-issue-template-client-iui0o
  - server: https://codesandbox.io/s/urql-issue-template-server-0ufyz
-->

--- a/.changeset/old-colts-breathe.md
+++ b/.changeset/old-colts-breathe.md
@@ -1,0 +1,5 @@
+---
+'@urql/exchange-graphcache': patch
+---
+
+Loosen type constraint on `ScalarObject` to account for custom scalar deserialization like `Date` for `DateTime`s.

--- a/exchanges/graphcache/src/types.ts
+++ b/exchanges/graphcache/src/types.ts
@@ -13,7 +13,8 @@ export interface Fragments {
 export type Primitive = null | number | boolean | string;
 
 export interface ScalarObject {
-  [key: string]: unknown;
+  constructor?: Function;
+  [key: string]: any;
 }
 
 export type Scalar = Primitive | ScalarObject;

